File: update-builder/src/main/java/de/darkatra/patcher/updatebuilder/UpdateBuilderNoUI.java
package de.darkatra.patcher.updatebuilder;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import de.darkatra.patcher.updatebuilder.service.model.Compression;
import de.darkatra.patcher.updatebuilder.service.model.ObsoleteFile;
import de.darkatra.patcher.updatebuilder.service.model.Packet;
import de.darkatra.patcher.updatebuilder.service.model.Patch;
import lombok.RequiredArgsConstructor;
import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Instant;
import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.zip.GZIPOutputStream;

public class UpdateBuilderNoUI {

	@RequiredArgsConstructor
	enum Directory {
		/**
		 * For example: C:\Users\<User>\AppData\Roaming\Meine Der Herr der Ringe™, Aufstieg des Hexenkönigs™-Dateien\
		 */
		APPDATA_DIR_NAME("appdata", "${rotwkUserDir}"),
		/**
		 * B:\Electronic Arts\Die Schlacht um Mittelerde II\
		 */
		BMFE2_DIR_NAME("bfme2", "${bfme2HomeDir}"),
		/**
		 * B:\Electronic Arts\Aufstieg des Hexenkönigs\
		 */
		ROTWK_DIR_NAME("rotwk", "${rotwkHomeDir}"),
		/**
		 * C:\Users\<User>\AppData\Roaming\.patcher\
		 */
		PATCH_DIR_NAME("patcher", "${patcherUserDir}");

		private final String name;
		private final String contextVariable;
	}

	private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
		.findAndRegisterModules()
		.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
	private static final Path OBSOLETE_FILES_PATH = Path.of("./obsolete-files.json");
	private static final Set<String> FILES_REQUIRING_BACKUP = Set.of("asset.dat", "game.dat");
	private static final HashingService HASHING_SERVICE = new HashingService();

	public static void main(final String[] args) throws IOException, InterruptedException {

		final Patch patch = new Patch();

		final ObsoleteFile[] obsoleteFiles = OBJECT_MAPPER.readValue(OBSOLETE_FILES_PATH.toFile(), ObsoleteFile[].class);
		patch.setObsoleteFiles(Arrays.stream(obsoleteFiles).collect(Collectors.toSet()));

		for (final Directory directory : Directory.values()) {
			final Path basePath = Path.of("./" + directory.name);
			for (final Path filePath : readFilesInDirectory(basePath)) {
				addFilesToPatch(patch, directory, filePath);
			}
		}

		Files.writeString(
			Path.of("./output/version.json"),
			OBJECT_MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(patch),
			StandardOpenOption.WRITE,
			StandardOpenOption.TRUNCATE_EXISTING,
			StandardOpenOption.CREATE
		);
	}

	private static void addFilesToPatch(final Patch patch, final Directory directory, final Path filePath) throws IOException, InterruptedException {

		final File gzipFile = createGzipArchive(directory, filePath);

		patch.getPackets().add(
			new Packet()
				.setSrc(Path.of("${serverUrl}/bfmemod2/").resolve(
					Path.of(directory.name).relativize(Path.of(filePath.toString() + ".gz"))
				).normalize().toString().replace("\\", "/"))
				.setDest(Path.of(directory.contextVariable).resolve(
					Path.of(directory.name).relativize(filePath)
				).normalize().toString().replace("\\", "/"))
				.setPacketSize(filePath.toFile().length())
				.setDateTime(Instant.now())
				.setChecksum(HASHING_SERVICE.getSHA3Checksum(filePath.toFile())
					.orElseThrow(() -> new IllegalStateException("Could not calculate the hash for: " + filePath.toFile())))
				.setBackupExisting(FILES_REQUIRING_BACKUP.contains(filePath.toFile().getName()))
				.setCompression(Compression.ZIP)
		);
	}

	private static File createGzipArchive(final Directory directory, final Path input) throws IOException {

		final File outputFile = new File(Path.of("./output/", Path.of(directory.name).relativize(input).toString()).normalize().toString() + ".gz");
		outputFile.getParentFile().mkdirs();
		try (final FileInputStream inputStream = new FileInputStream(input.toFile());
			 final GZIPOutputStream outputStream = new GZIPOutputStream(new FileOutputStream(outputFile))) {
			IOUtils.copy(inputStream, outputStream);
		}
		return outputFile;
	}

	private static Set<Path> readFilesInDirectory(final Path directory) throws IOException {
		try (Stream<Path> stream = Files.walk(directory)) {
			return stream
				.filter(Files::isRegularFile)
				.collect(Collectors.toSet());
		}
	}
}

File: update-builder/src/main/kotlin/de/darkatra/patcher/updatebuilder/UpdateBuilderNoUI.kt
package de.darkatra.patcher.updatebuilder

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonMapperBuilder
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.time.Instant
import java.util.stream.Collectors
import java.util.zip.GZIPOutputStream
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi
import kotlin.io.path.createParentDirectories
import kotlin.io.path.exists
import kotlin.io.path.fileSize
import kotlin.io.path.inputStream
import kotlin.io.path.outputStream
import kotlin.io.path.pathString

fun main() {
    UpdateBuilderNoUI.build()
}

object UpdateBuilderNoUI {

    private val objectMapper: ObjectMapper = jacksonMapperBuilder().addModule(JavaTimeModule()).build()
    private val obsoleteFilesPath = Path.of("./obsolete-files.json")
    private val featuresPath = Path.of("./features.json")
    private val filesRequireBackup = setOf("asset.dat", "game.dat")
    private val hashingService = HashingService

    fun build() {

        val startTime = System.nanoTime()

        println("Creating new patch...")

        println("Reading version.json...")
        val versionJsonPath = Path.of("./output/version.json")
        val lastPatch: Patch? = when {
            versionJsonPath.toFile().exists() -> objectMapper.readValue(versionJsonPath.toFile(), Patch::class.java)
            else -> null
        }

        println("Applying obsolete files from: ${obsoleteFilesPath.toFile().path}")
        val obsoleteFiles: Set<ObsoleteFile> = objectMapper.readValue(obsoleteFilesPath.toFile(), Array<ObsoleteFile>::class.java).toSet()

        println("Applying features from: ${featuresPath.toFile().path}")
        val featureFiles: Set<FeatureFile> = objectMapper.readValue(featuresPath.toFile(), Array<FeatureFile>::class.java).toSet()

        var added = 0
        var archived = 0
        val packets = mutableSetOf<Packet>()
        for (directory in Directory.entries) {
            val basePath = Path.of("./" + directory.dirName)
            for (filePath in readFilesInDirectory(basePath)) {
                println("* Adding file: ${filePath.pathString}")
                if (addFilesToPatch(packets, lastPatch, directory, filePath, featureFiles)) {
                    archived++
                }
                added++
            }
        }
        println("Files added: $added, Files archived: $archived")

        val patch = Patch(
            obsoleteFiles = obsoleteFiles,
            packets = packets.toSet()
        )

        println("Deleting obsolete files from output folder...")

        var deleted = 0
        readFilesInDirectory(Path.of("./output")).forEach { distributionFile ->
            if (patch.packets.none { packet -> Files.isSameFile(packet.gzipPath!!, distributionFile) }) {
                println("* Deleting file: ${distributionFile.pathString}")
                Files.delete(distributionFile)
                deleted++
            }
        }

        println("Removing empty folders...")

        var deletedEmptyFolders = 0
        readFoldersInDirectory(Path.of("./output"))
            .reversed()
            .filter { folder -> folder.toFile().list()?.isEmpty() ?: false }
            .forEach { folder ->
                deletedEmptyFolders += deleteFolder(Path.of("./output"), folder)
            }
        println("Files deleted: $deleted, empty Folders deleted: $deletedEmptyFolders")

        println("Writing version.json...")
        Files.writeString(
            versionJsonPath,
            objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(patch),
            StandardOpenOption.WRITE,
            StandardOpenOption.TRUNCATE_EXISTING,
            StandardOpenOption.CREATE
        )

        val endTime = System.nanoTime()
        println("Success! Took: ${(endTime - startTime) / 1000000}ms")
    }

    @OptIn(ExperimentalEncodingApi::class)
    private fun addFilesToPatch(packets: MutableSet<Packet>, lastPatch: Patch?, directory: Directory, filePath: Path, featureFiles: Set<FeatureFile>): Boolean {

        // creating the archive takes a lot of time, check if it's really necessary by comparing checksums
        // this only compares the checksum of the source file and will not ensure that the archive is identical!
        // manually deleting the version.json will always trigger the creation of a new archive
        val checksum = hashingService.getSHA3Checksum(filePath.toFile())
            .orElseThrow { IllegalStateException("Could not calculate the hash for: ${filePath.toFile()}") }

        val dest = Path.of(directory.contextVariable).resolve(
            Path.of(directory.dirName).relativize(filePath)
        ).normalize().toString().replace("\\", "/")

        var fileChanged = true
        if (lastPatch != null) {
            val lastPatchChecksum: String? = lastPatch.packets.firstOrNull { p -> p.dest == dest }?.checksum
            if (lastPatchChecksum != null && checksum == lastPatchChecksum) {
                fileChanged = false
            }
        }

        val base64EncodedFilePath = Base64.UrlSafe.encode(Path.of(directory.dirName).relativize(filePath).normalize().toString().toByteArray()) + ".gz"
        val output = Path.of("./output/", base64EncodedFilePath)
        val outputExists: Boolean = output.exists()
        val archive: Boolean = fileChanged || !outputExists
        if (archive) {
            println("** Archiving file (changed: $fileChanged, exists: $outputExists): ${output.pathString}")
            createGzipArchive(filePath, output)
        }

        val feature = featureFiles.find { featureFile -> featureFile.dest == dest }?.feature
        packets.add(
            Packet(
                src = Path.of("\${serverUrl}/bfmemod2/").resolve(base64EncodedFilePath).normalize().toString().replace("\\", "/"),
                dest = dest,
                packetSize = filePath.toFile().length(),
                compressedSize = output.fileSize(),
                dateTime = Instant.now(),
                checksum = checksum,
                backupExisting = filesRequireBackup.contains(filePath.toFile().name),
                compression = Compression.ZIP,
                feature = feature,
                gzipPath = output
            )
        )

        return archive
    }

    private fun createGzipArchive(input: Path, output: Path) {
        output.createParentDirectories()
        input.inputStream().use { inputStream ->
            GZIPOutputStream(output.outputStream()).use { outputStream ->
                inputStream.transferTo(outputStream)
            }
        }
    }

    private fun readFilesInDirectory(directory: Path): Set<Path> {
        return when (directory.exists()) {
            true -> Files.walk(directory).use { stream ->
                stream
                    .filter { path: Path -> Files.isRegularFile(path) }
                    .collect(Collectors.toSet())
            }

            else -> emptySet()
        }
    }

    private fun readFoldersInDirectory(directory: Path): Set<Path> {
        return when (directory.exists()) {
            true -> Files.walk(directory).use { stream ->
                stream
                    .filter { path: Path -> Files.isDirectory(path) }
                    .collect(Collectors.toSet())
            }

            else -> emptySet()
        }
    }

    private fun deleteFolder(topLevelDir: Path, currentPath: Path): Int {
        var deleted = 0
        if (topLevelDir != currentPath && currentPath.startsWith(topLevelDir)) {
            val parent = currentPath.parent
            Files.delete(currentPath)
            deleted += 1 + if (!Files.list(parent).findAny().isPresent) {
                deleteFolder(topLevelDir, parent)
            } else {
                0
            }
        }
        return deleted
    }
}
